Sources/Datadog/Core/Utils/ValuePublisher.swift

```diff
@@ -0,0 +1,110 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-2020 Datadog, Inc.
+ */
+
+import Foundation
+
+/// An observer subscribing to a `ValuePublisher`.
+internal protocol ValueObserver {
+    associatedtype ObservedValue
+
+    /// Notifies this observer on the value change. Called on the publisher's queue.
+    /// If the `ObservedValue` conforms to `Equatable`, only distinct changes will be notified.
+    func onValueChanged(oldValue: ObservedValue, newValue: ObservedValue)
+}
+
+/// Manages the `Value` in a thread safe manner and notifies subscribed `ValueObservers` on its change.
+internal class ValuePublisher<Value> {
+    /// Type erasure for `ValueObserver` type.
+    private struct AnyObserver<ObservedValue> {
+        let notifyValueChanged: (ObservedValue, ObservedValue) -> Void
+
+        init<Observer: ValueObserver>(wrapped: Observer) where Observer.ObservedValue == ObservedValue {
+            self.notifyValueChanged = wrapped.onValueChanged
+        }
+    }
+
+    /// The queue used to synchronize the access to the `unsafeValue`.
+    /// Concurrent queue is used for performant reads, `.barrier` must be used to make writes exclusive.
+    private let concurrentQueue = DispatchQueue(
```

Comment:
I used `concurrentQueue` on purpose, hoping that it will make the next maintainer think twice before typing `.sync {}` or `.async {}`, as now the `.barrier` flag is very important.

Reply:
That makes sense, let's keep `concurrentQueue` then 👍